codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/OperationGenerator.java

```diff
@@ -0,0 +1,123 @@
+
+    private final GoSettings settings;
+    private final Model model;
+    private final SymbolProvider symbolProvider;
+    private final GoWriter writer;
+    private final ServiceShape service;
+    private final OperationShape operation;
+    private final Symbol operationSymbol;
+    private final ApplicationProtocol applicationProtocol;
+
+    OperationGenerator(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            ServiceShape service,
+            OperationShape operation,
+            Symbol operationSymbol,
+            ApplicationProtocol applicationProtocol
+    ) {
+        this.settings = settings;
+        this.model = model;
+        this.symbolProvider = symbolProvider;
+        this.writer = writer;
+        this.service = service;
+        this.operation = operation;
+        this.operationSymbol = operationSymbol;
+        this.applicationProtocol = applicationProtocol;
+    }
+
+    @Override
+    public void run() {
+        OperationIndex operationIndex = model.getKnowledge(OperationIndex.class);
+        Symbol serviceSymbol = symbolProvider.toSymbol(service);
+
+        if (!operationIndex.getInput(operation).isPresent()) {
+            // Theoretically this shouldn't ever get hit since we automatically insert synthetic inputs / outputs.
+            throw new CodegenException(
+                    "Operations are required to have input shapes in order to allow for future evolution.");
+        }
+        StructureShape inputShape = operationIndex.getInput(operation).get();
+        Symbol inputSymbol = symbolProvider.toSymbol(inputShape);
+
+        if (!operationIndex.getOutput(operation).isPresent()) {
+            throw new CodegenException(
+                    "Operations are required to have output shapes in order to allow for future evolution.");
+        }
+        StructureShape outputShape = operationIndex.getOutput(operation).get();
+        Symbol outputSymbol = symbolProvider.toSymbol(outputShape);
+
+        writer.writeShapeDocs(operation);
+        Symbol contextSymbol = SymbolUtils.createValueSymbolBuilder("Context", GoDependency.CONTEXT).build();
+        writer.openBlock("func (c $P) $T(ctx $T, params $P, opts ...func(*Options)) ($P, error) {", "}",
+                serviceSymbol, operationSymbol, contextSymbol, inputSymbol, outputSymbol, () -> {
+                    constructStack();
+
+                    writer.write("options := c.options.Copy()");
+                    writer.openBlock("for _, fn := range optFns {", "}", () -> {
+                        writer.write("fn(&options)");
+                    });
+                    writer.openBlock("for _, fn := range options.APIOptions {", "}", () -> {
+                        writer.write("if err := fn(stack); err != nil { return nil, err }");
+                    });
+
+                    writer.write("result, err := handler.Handle(ctx, params)");
```

Comment:
Nit handler will return , `result, metadata, err`.  for the return can do the following to capture the metadata in the result output shape.

```
    out := res.($P)
    out.ResultMetadata = metadata
    
    return out, nil
```